File: web/src/entities/spawnPoint.ts
import { ImmutableVec2 } from '../math.js';
import { Entity } from '../entity.js';
import { PlayerSkill } from '../playerSkill.js';
import { Resource } from '../resource.js';
import type { Animation } from '../media/animation.js';

const enum SpawnState { UNCHECKED = 0, SEMI_CHECKED = 1, CHECKED = 2 }

export class SpawnPoint extends Entity {
  private state: SpawnState = SpawnState.UNCHECKED;
  private frame = 0;
  private readonly animation: Animation;

  constructor() {
    super();
    this.setSize(new ImmutableVec2(22, 25));
    this.animation = Resource.getAnimation('data/images/checkpoint.bmp', 3);
  }

  getLayer(): number { return 3; }

  override draw(ctx: CanvasRenderingContext2D, offsetX: number, offsetY: number, _layer: number): void {
    const x = Math.round(this.getPosition().x - this.animation.getFrameWidth() / 2 + offsetX);
    const y = Math.round(this.getPosition().y - this.animation.getFrameHeight() / 2 + offsetY);
    this.animation.drawFrame(ctx, this.state, x, y);
  }

  override update(): void {
    const hero = this.myRoom.getHero();
    if (hero.Collides(this)) {
      if (this.state === SpawnState.UNCHECKED) {
        PlayerSkill.playerDidSomethingClever(0.7, 0.5);
      }
      this.state = SpawnState.SEMI_CHECKED;
      hero.setLastSpawnPoint(this.getPosition());
    }

    if (this.state === SpawnState.SEMI_CHECKED) this.frame++;
    if (this.frame > 5) this.state = SpawnState.CHECKED;
  }
}

File: wii/src/entities/spawn_point.cpp
#include "entities/spawn_point.h"
#include "entities/hero.h"
#include "resource.h"
#include "media/animation.h"
#include "room.h"
#include "player_skill.h"
#include <cmath>

SpawnPoint::SpawnPoint() {
    setSize(Vec2(22, 25));
    animation_ = Resource::getAnimation("data/images/checkpoint.bmp", 3);
}

void SpawnPoint::draw(SDL_Renderer* renderer, int offsetX, int offsetY, int /*layer*/) {
    int x = static_cast<int>(std::round(position_.x)) - animation_->getFrameWidth() / 2 + offsetX;
    int y = static_cast<int>(std::round(position_.y)) - animation_->getFrameHeight() / 2 + offsetY;
    animation_->drawFrame(renderer, static_cast<int>(state_), x, y);
}

void SpawnPoint::update() {
    Hero* hero = room_->getHero();
    if (hero->Collides(*this)) {
        if (state_ == SpawnState::UNCHECKED) {
            PlayerSkill::playerDidSomethingClever(0.7f, 0.5f);
        }
        state_ = SpawnState::SEMI_CHECKED;
        hero->setLastSpawnPoint(position_);
    }

    if (state_ == SpawnState::SEMI_CHECKED) frame_++;
    if (frame_ > 5) state_ = SpawnState::CHECKED;
}
